charts/consul/templates/gossip-encryption-autogeneration-job.yaml

```diff
@@ -0,0 +1,76 @@
+{{- if .Values.global.gossipEncryption.autoGenerate }}
+{{- if (or .Values.global.gossipEncryption.secretName .Values.global.gossipEncryption.secretKey) }}
+  {{ fail "If global.gossipEncryption.autoGenerate is true, global.gossipEncryption.secretName and global.gossipEncryption.secretKey must not be set." }}
+{{ end }}
+---
+# automatically generate encryption key for gossip protocol and save it in Kubernetes secret
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ template "consul.fullname" . }}-gossip-encryption-autogenerate
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: {{ template "consul.name" . }}
+    chart: {{ template "consul.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  annotations:
+    "helm.sh/hook": pre-install,pre-upgrade
+    "helm.sh/hook-weight": "1"
+    "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
+spec:
+  template:
+    metadata:
+      name: {{ template "consul.fullname" . }}-gossip-encryption-autogenerate
+      labels:
+        app: {{ template "consul.name" . }}
+        chart: {{ template "consul.chart" . }}
+        release: {{ .Release.Name }}
+        component: gossip-encryption-autogen
+      annotations:
+        "consul.hashicorp.com/connect-inject": "false"
+    spec:
+      restartPolicy: Never
+      serviceAccountName: {{ template "consul.fullname" . }}-gossip-encryption-autogenerate
+      containers:
+        - name: gossip-encryption-autogen
+          image: "{{ .Values.global.image }}"
+          env:
+            - name: NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+          workingDir: /tmp
+          command:
+            - "/bin/sh"
+            - "-ec"
+            - |
+              secretName={{ template "consul.fullname" . }}-gossip-encryption-key
+              secretKey=key
+              # Check if secret already exists
+              currentGossipSecret=$(curl -s -X GET --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
+                https://${KUBERNETES_SERVICE_HOST}:${KUBERNETES_SERVICE_PORT}/api/v1/namespaces/${NAMESPACE}/secrets/${secretName} \
+                -H "Authorization: Bearer $( cat /var/run/secrets/kubernetes.io/serviceaccount/token )" \
+                -H "Content-Type: application/json" \
+                -H "Accept: application/json")
+              doesCurrentSecretNameExist=$(echo $currentGossipSecret | jq '.code==200')
+              if [ "$doesCurrentSecretNameExist" = "true" ]; then
+                doesCurrentSecretKeyExist=$(echo $currentGossipSecret | jq '.data.key')
+              fi
```

Comment:
Can you elaborate on the scenario here? If I understand this correctly, this is a case where the secret exists, but the secret does not have a value set on the key `key`?  I'm a touch hesitant about modifying an existing secret that wasn't created by this job.

Additionally, is there a way for us to write a BATS test for this?

Reply:
Yeah I think we need to exit 0 in this case since we don't want to modify an existing secret.

I think we can also use a POST request and ignore secrets already exists response code. I think it'll make the logic a bit simpler. It's what we used to do in the tls-init job: https://github.com/hashicorp/consul-helm/blob/c0722c63907fe22ca92f7b021caf0647de8c0028/templates/tls-init-job.yaml#L55-L58

Reply:
I'll try to replicate the tls code. That is helpful. Then I don't need to do all the checks. I'll reach out after I implement it that way to make sure I did it right.

Reply:
This works now by just POSTing, but it won't overwrite the secret if it's already there.